### ISSUE TITLE
Changing syncExpression does not update the DeltaSync 

### ISSUE THREAD

**Issue by dgagnon** [aws-amplify/amplify-flutter#3703]
### Description

To avoid the long initial load time as mentioned here, we tried a two-step approach where we have an initial and a default datastore configuration.  Like most users, we have a User model that needs to be handled first and early, so like most, we wait until the datastore is fully ready before accessing it.  Then we reset the syncExpression and restart the datastore to get the full load, allowing a faster loading of the user's preference and interactiveness. 

Unfortunately, all my efforts to have the syncExpression re-evaluated are failing on android.  I can inspect the datastore plugin config and see the the syncExpressions have changed, but it just does not sync the new data.  I have tried so many things that the code I have right now is probably not ideal, but here is part of the class anyway.  The ready first event triggers the full Load and then the second ready events actually returns control to the app.

Originally from #1479 .

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [X] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Steps to Reproduce

1. Setup amplify and the datastore
2. Insert over 10k items in the datastore
3. Start the App, notice it takes many minutes before datastore is ready.
4. Clear the datastore
5. Enable two-stage loading
6. Notice it is not working

### Screenshots

```dart
  // How can we set this to false if the datastore is already fully initialized ? do a query ?
  bool isInitialSync = true;

  // default is 86400
  int initialSyncInterval = 8640000;
  int defaultSyncInterval = 864000;

  int? get syncInterval {
    return isInitialSync ? initialSyncInterval : defaultSyncInterval;
  }

  set syncInterval(int? value) {
    datastorePlugin.syncInterval = value ?? defaultSyncInterval;
  }

  // default is 10000
  int initialSyncMaxRecords = 100;
  int defaultSyncMaxRecords = 100000;

  int? get syncMaxRecords {
    return isInitialSync ? initialSyncMaxRecords : defaultSyncMaxRecords;
  }

  set syncMaxRecords(int? value) {
    datastorePlugin.syncMaxRecords = value ?? defaultSyncMaxRecords;
  }

  // default is 1000
  int initialSyncPageSize = 100;
  int defaultSyncPageSize = 1000;

  int? get syncPageSize {
    return isInitialSync ? initialSyncPageSize : defaultSyncPageSize;
  }

  set syncPageSize(int? value) {
    datastorePlugin.syncPageSize = value ?? defaultSyncPageSize;
  }

  QueryPredicate<Model> initialUserSyncExpression = User.STATUS.eq(UserStatus.ACTIVE).or(User.STATUS.eq(UserStatus.SUSPENDED));
  QueryPredicate<Model> defaultUserSyncExpression = User.STATUS.eq(UserStatus.ACTIVE).or(User.STATUS.eq(UserStatus.SUSPENDED));
  QueryPredicate<Model> getUserSyncExpression() {
    return isInitialSync ? initialUserSyncExpression : defaultUserSyncExpression;
  }

  QueryPredicate<Model> initialProjectSyncExpression = Project.ID.eq("1");
  QueryPredicate<Model> defaultProjectSyncExpression = Project.STATUS.eq(Status.ACTIVE).or(Project.STATUS.eq(Status.COMPLETED));
  QueryPredicate<Model> getProjectSyncExpression() {
    return isInitialSync ? initialProjectSyncExpression : defaultProjectSyncExpression;
  }

  QueryPredicate<Model> initialTaskSyncExpression = Task.ID.eq("1");
  QueryPredicate<Model> defaultTaskSyncExpression = Task.STATUS.eq(Status.ACTIVE).or(Project.STATUS.eq(Status.COMPLETED));
  QueryPredicate<Model> getTaskSyncExpression() {
    return isInitialSync ? initialTaskSyncExpression : defaultTaskSyncExpression;
  }

  late List<DataStoreSyncExpression> syncExpressions = [
    DataStoreSyncExpression(
      // should query by ID directly since we need our user, since we limit the amount of records, an account with lots of shared User could possibly not received its own user this way
      User.classType,
      getUserSyncExpression,
    ),
    DataStoreSyncExpression(
      Project.classType,
      getProjectSyncExpression,
    ),
    DataStoreSyncExpression(
      Task.classType,
      getTaskSyncExpression,
    )
  ];

  /// When there are lots of items, the datastore takes forever to load, even on a good connection.
  /// By restricting whats get synced on an initial load (User profile and anything else needed to get started)
  /// We can return interactiveness to the user faster.
  Future<void> loadFullDatastore() async {
    isInitialSync = false;
    datastorePlugin.syncExpressions = syncExpressions;
    syncMaxRecords = defaultSyncMaxRecords;
    syncInterval = defaultSyncInterval;
    syncPageSize = defaultSyncPageSize;
    dbg("resetting Datastore", cat: 'DATASTORE');

    // using Timer to avoid all sorts of weird timing issues with the underlying libraries.
    // it was actually crashing the app.  these timings worked on the first try, did not try other ones.
    Timer(const Duration(seconds: 1), () async {
      try {
        await Amplify.DataStore.stop();
        Timer(const Duration(seconds: 3), () async {
          try {
            await Amplify.DataStore.start();
          } on Exception catch (error, stackTrace) {
            err('Error starting DataStore: $error', stackTrace, cat: 'DATASTORE');
            if (kDebugMode) rethrow;
          }
        });
      } catch (error, stackTrace) {
        err('Error stopping DataStore: $error', stackTrace, cat: 'DATASTORE');
        if (kDebugMode) rethrow;
      }
    });
    final test = datastorePlugin.syncExpressions;
    dbg(test);
  }

  /// Configure Datastore to use queries instead of table scans.
  /// https://docs.amplify.aws/lib/datastore/sync/q/platform/flutter/#advanced-use-case---query-instead-of-scan
  /// And perform a two stage loading
  /// https://docs.amplify.aws/lib/datastore/conflict/q/platform/flutter/#custom-configuration
  late AmplifyDataStore datastorePlugin = AmplifyDataStore(
    modelProvider: ModelProvider.instance,

    /// Pools are default with iam secondary for admin access.
    authModeStrategy: AuthModeStrategy.multiAuth,

    /// Custom ErrorHandler to propagate errors to the user as needed
    errorHandler: ((error) {
      err('Custom DataStore ErrorHandler received: $error', null, cat: 'DATASTORE');
    }),

    /// Sync configuration defaults:
    syncInterval: syncInterval,
    syncMaxRecords: syncMaxRecords,
    syncPageSize: syncPageSize,

    // TODO: make the sync expression configurable
    syncExpressions: syncExpressions!,
  );

///// SNIP ///////

          switch (hubEvent.type) {
            /// We received a ready event, emitting [readyState]
            case DataStoreHubEventType.ready:
              // we should probably have 2 events this way the authService can handle the User and Profile setup  and the App can trigger on the real ready
              if (isInitialSync) {
                await loadFullDatastore();
              } else if (!isDatastoreReady) {
                isDatastoreReady = true;
                readyState.enter();
              }
...
```

**But the sync never changes:**  
[GETX] [DATASTORE] package:flutter_demo_with_docs/modules/aws/services/datastore.dart:330:15 <String> In DatastoreService._configureDatastoreService.<fn> => (Task)  [DeltaSync]. added 0, updated 0, deleted 0                    [DATASTORE]
  
**Expected result:**
[GETX] [DATASTORE] package:flutter_demo_with_docs/modules/aws/services/datastore.dart:219:15 <String> In DatastoreService._configureDatastoreService.<fn> => (Task)  [DeltaSync]. added 3244, updated 0, deleted 0                    [DATASTORE]

### Platforms

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.13.1

### Amplify Flutter Version

1.4.0

### Deployment Method

Amplify CLI

### Schema

```GraphQL
# all models in this schema. Learn more about authorization rules here: https://docs.amplify.aws/cli/graphql/authorization-rules
# TODO: look into making a UserInterface and a PPInterface: https://graphql.org/learn/schema/#interfaces
enum Status {
    ACTIVE
    COMPLETED
    DELETED
}

enum UserStatus {
    ACTIVE
    SUSPENDED
    DELETED
}

type User @model @auth(rules: [{ allow: owner, ownerField: "owner"}, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }]) {
    id: ID! @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }])
    username: String! @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }])
    data: AWSJSON @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }])
    attributes: AWSJSON @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }])
    # https://docs.amplify.aws/lib/datastore/sync/q/platform/flutter/#advanced-use-case---query-instead-of-scan
    status: UserStatus @default(value: "ACTIVE" ) @index(name: "byStatus") @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }])
    owner: String @index(name: "byOwner") @auth(rules: [{ allow: owner, ownerField: "owner", operations: [read, create, delete] }, { allow: private, provider: iam }, { allow: groups, groups: ["admin"] }])
    tasks: [Task] @hasMany(indexName: "byUser", fields: ["id"])
    sharedWith: [String] @auth(rules: [{ allow: owner, ownerField: "owner" }, { allow: private, provider: iam }, { allow: groups, groups: ["admin"] }])
}
type Project @model @auth(rules: [{ allow: owner, ownerField: "owner"}, {allow: private, operations: [read]}, {allow: private, provider: iam}]) {
    id: ID! @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, {allow: private, operations: [read]}, { allow: groups, groups: ["admin"] }])
    title: String! @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, {allow: private, operations: [read]}, { allow: groups, groups: ["admin"] }])
    description: String @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, {allow: private, operations: [read]}, { allow: groups, groups: ["admin"] }])
    data: AWSJSON @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, {allow: private, operations: [read]}, { allow: groups, groups: ["admin"] }])
    status: Status @default(value: "ACTIVE" ) @index(name: "byStatus") @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, {allow: private, operations: [read]}, { allow: groups, groups: ["admin"] }])
    dueAt: AWSDateTime! @auth(rules: [{ allow: owner, ownerField: "owner"}, { allow: private, provider: iam }, {allow: private, operations: [read]}, { allow: groups, groups: ["admin"] }])
    owner: String @index(name: "byOwner") @auth(rules: [{ allow: owner, ownerField: "owner", operations: [read, create, delete] }, { allow: private, provider: iam }, {allow: private, operations: [read]},{ allow: groups, groups: ["admin"] }])
    tasks: [Task] @hasMany(indexName: "byProject", fields: ["id"]) #https://docs.amplify.aws/cli/graphql/data-modeling/#belongs-to-relationship
}
type Task @model @auth(rules: [{ allow: owner, ownerField: "owner"}, {allow: private, provider: iam}, { allow: groups, groups: ["admin"], operations: [read] }]) {
    id: ID @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"], operations: [read] }])
    title: String! @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"], operations: [read] }])
    description: String @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"], operations: [read] }])
    data: AWSJSON @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"], operations: [read] }])
    status: Status @default(value: "ACTIVE" ) @index(name: "byStatus") @auth(rules: [{ allow: owner, ownerField: "owner" }, {allow: private, provider: iam}, { allow: groups, groups: ["admin"] }])
    dueAt: AWSDateTime! @auth(rules: [{ allow: owner, ownerField: "owner" }, { allow: private, provider: iam }, { allow: groups, groups: ["admin"], operations: [read] }])
    # https://docs.amplify.aws/cli/graphql/authorization-rules/#multi-user-data-access
    #    owners: [String]
    # BUG: https://github.com/aws-amplify/amplify-flutter/issues/1566
    owner: String @index(name: "byOwner") @auth(rules: [{ allow: owner, ownerField: "owner", operations: [read, create, delete] }, { allow: private, provider: iam }, { allow: groups, groups: ["admin"], operations: [read] }])
    projectID: ID! @index(name: "byProject") #https://docs.amplify.aws/cli/graphql/data-modeling/#belongs-to-relationship
    project: Project! @belongsTo(fields: ["projectID"])
    sharedWith: [String] @auth(rules: [{ allow: owner, ownerField: "owner" }, { allow: private, provider: iam }, { allow: groups, groups: ["admin"], operations: [read] }])
    userID: ID @index(name: "byUser") #https://docs.amplify.aws/cli/graphql/data-modeling/#belongs-to-relationship
    user: User @belongsTo(fields: ["userID"])
}
# user should not be nullable, but I didnt want to wipe the db
```

**Comment by Jordan-Nelson** [comment]
Hello @dgagnon - I am unable to reproduce this. I have created a sample app, added over 1,000 models, and I am seeing that changing the sync expression is working as expected. I will share the details of the sample app and repro steps below. If you are still facing this issue would you be able to try to create a minimal reproduction with specific steps?

Here is the schema/app I used and the steps I took.

Schema:

<details>
```graphql
# This "input" configures a global authorization rule to enable public access to
# all models in this schema. Learn more about authorization rules here: https://docs.amplify.aws/cli/graphql/authorization-rules
input AMPLIFY {
  globalAuthRule: AuthRule = { allow: public }
} # FOR TESTING ONLY!
type Todo @model {
  id: ID!
  name: String!
  description: String
}
```

</details>

App: 

<details>

```dart
import 'dart:async';

import 'package:amplify_api/amplify_api.dart';
import 'package:amplify_datastore/amplify_datastore.dart';
import 'package:amplify_flutter/amplify_flutter.dart';
import 'package:flutter/material.dart';
import 'package:issue_3703/amplifyconfiguration.dart';
import 'package:issue_3703/models/ModelProvider.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: true),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final formKey = GlobalKey<FormState>();

  List<String> todoNames = [];
  List<Todo> todos = [];
  StreamSubscription<QuerySnapshot<Todo>>? sub;
  void subscribeToTodos() {
    sub?.cancel();
    sub = Amplify.DataStore.observeQuery(Todo.classType, sortBy: [
      Todo.NAME.ascending(),
    ]).listen((event) {
      setState(() {
        todos = event.items;
      });
    });
  }

  Future<void> _configureAmplify() async {
    try {
      await Amplify.addPlugin(AmplifyAPI());
      await Amplify.addPlugin(
        AmplifyDataStore(
          modelProvider: ModelProvider.instance,
          syncExpressions: [
            DataStoreSyncExpression(Todo.classType, () {
              if (todoNames.isEmpty) {
                return QueryPredicate.all;
              }
              return Todo.NAME.isOneOf(todoNames);
            })
          ],
          syncMaxRecords: 100,
        ),
      );
      await Amplify.configure(amplifyconfig);
      safePrint('Successfully configured');
      subscribeToTodos();
    } on Exception catch (e) {
      safePrint('Error configuring Amplify: $e');
    }
  }

  @override
  void initState() {
    _configureAmplify();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
      ),
      persistentFooterButtons: [
        OutlinedButton(
          onPressed: () async {
            for (final item in List.generate(
              100,
              (index) => Todo(
                name: 'todo $index',
                description: UUID.getUUID(),
              ),
            )) {
              await Amplify.DataStore.save(item);
            }
          },
          child: const Text('Create 100 todos'),
        ),
        SizedBox(width: 12),
        OutlinedButton(
          onPressed: () async {
            await Amplify.DataStore.clear();
            subscribeToTodos();
            await Amplify.DataStore.start();
          },
          child: const Text('Resync Data'),
        )
      ],
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Form(
                key: formKey,
                child: Row(
                  children: [
                    Expanded(
                      child: TextFormField(
                        decoration: const InputDecoration(
                          label: Text('Sync Expression Todo name'),
                        ),
                        onSaved: (value) {
                          if (value == null || value.isEmpty) return;
                          setState(() {
                            todoNames.add(value);
                            formKey.currentState?.reset();
                          });
                        },
                      ),
                    ),
                    FilledButton(
                      onPressed: () => formKey.currentState?.save(),
                      child: const Text("Add"),
                    )
                  ],
                ),
              ),
            ),
            Row(
              children: [
                for (final name in todoNames)
                  Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: Chip(
                      label: Text(name),
                      onDeleted: () {
                        setState(() {
                          todoNames.remove(name);
                        });
                      },
                    ),
                  ),
              ],
            ),
            Expanded(
              child: todos.isEmpty
                  ? const Center(child: CircularProgressIndicator())
                  : ListView(
                      children: [
                        for (final todo in todos.take(100))
                          ListTile(
                            title: Text(todo.name),
                            subtitle: Text(todo.description ?? ''),
                          ),
                      ],
                    ),
            )
          ],
        ),
      ),
    );
  }
}

extension QueryFieldX<T> on QueryField<T> {
  QueryPredicate isOneOf(List<T> items) {
    assert(items.isNotEmpty);
    if (items.length == 1) {
      return eq(items[0]);
    }
    QueryPredicateGroup queryPredicate = eq(items[0]).or(eq(items[1]));
    for (var i = 2; i < items.length; i++) {
      queryPredicate = queryPredicate.or(eq(items[i]));
    }
    return queryPredicate;
  }
}
```

</details>

Steps:
- Create over 1000 Todo models - 10 with the name "todo 0", 10 with the name "todo 1", etc. (tap create the button "100 todos" 10 times)
- Reload data, confirm data is loading
- Add a sync expression, reload data, confirm appropriate data is loading
- Update sync expression, reload data, confirm appropriate data is loading
- Remove sync expression, reload data, confirm appropriate data is loading

Below is a video if my attempt to reproduce this:

https://github.com/aws-amplify/amplify-flutter/assets/20613561/8e30ed1f-1367-4850-a3e0-19db211d80e8

**Comment by dgagnon** [comment]
Hi,
thanks for the reply.  I can only find one exemple of DataStoreSyncExpression in your code exemple and it contains the following code.
```dart
[
            DataStoreSyncExpression(Todo.classType, () {
              if (todoNames.isEmpty) {
                return QueryPredicate.all;
              }
              return Todo.NAME.isOneOf(todoNames);
            })
          ]
```  
This is different than what I am doing.  The difference is we are completely replacing the SyncExpression, with the getter being outside of it.  Here is an exemple of how different the syncExpressions are.
```dart
 QueryPredicate<Model> initialTaskSyncExpression = Task.ID.eq("1");
  QueryPredicate<Model> defaultTaskSyncExpression = Task.STATUS.eq(Status.ACTIVE).or(Project.STATUS.eq(Status.COMPLETED));
```

I will try to move the logic completely inside the () {} for each syncExpression to see if that resolves it.

**Comment by Lance-Kraakman** [comment]
I am facing a similar issue, I notice both my and @dgagnon have secondary indexes in our models (@index fields in our schema) @Jordan-Nelson maybe this could be the difference? 
(Here is my model)
![image](https://github.com/aws-amplify/amplify-flutter/assets/44858574/8234d7c1-806d-48b9-b931-bcf1f3060d21)

**Comment by Jordan-Nelson** [comment]
Apologies again the delay.

I have spent a bit of time trying to reproduce this with a schema that uses an index and have been unable to do so. The schema I am using is below.

```graphql
type Todo @model {
  id: ID!
  date: AWSDateTime @index(name: "byDate")
  name: String! @index(name: "byName")
  description: String
  status: Status @index(name: "byStatus")
}

enum Status {
  ACTIVE
  COMPLETED
  DELETED
}
```

I have attempted to reproduce this by setting selective sync expressions on each of the fields with indexes in the schema. In each case changing the sync expression at runtime works as expected.

If either of you can create minimal reproduction we can investigate further. A minimal reproduction should include:
- The schema (it can be the full schema you are using in your app)
- A sample app with the selective sync expression and mechanism to change it at runtime
- A description of any data required to populate the remote DB with in order to repro.

**Comment by dgagnon** [comment]
@Jordan-Nelson thank you for investigating.  Unfortunally, I was unable to create a small, reproducible demo and there are many parts to take into account here, including logging in to amplify.  I am unsure how to create such a demo and share it.  

I do not believe the issue is linked to indexes, but more to the way the query is changed.  You use a basic query, I change a basic query into a complex one.  It is not just variables being re-evaluated, but a while new function.

**Comment by Jordan-Nelson** [comment]
Thanks for the update @dgagnon 

@Lance-Kraakman - Let me know if you can provide a minimal reproduction. Otherwise I think we will close this for the time being and we can reopen when we have a reproduction to investigate.

**Comment by Jordan-Nelson** [comment]
Hi folks - I am going to close this out since we have been unable to reproduce this issue. If anyone can provide a minimal reproduction we can reopen this issue. Thanks.

**Comment by HarkerTech** [comment]
Hi,

We have encountered this issue in amplify_datastore v2.4.1 when trying to limit the initial sync and then expanding the query to match more results later. In our testing we have found that the query is reevaluated correctly when calling `Amplify.DataStore.clear` but not with `Amplify.DataStore.stop`.

Therefore a potential reason this could not be reproduced is that the attempted reproduction code uses clear whereas the original ticket example code uses stop.
